stdlib/LinearAlgebra/src/dense.jl

```diff
@@ -514,12 +514,38 @@ function (^)(A::AbstractMatrix{T}, p::Integer) where T<:Integer
 end
 function integerpow(A::AbstractMatrix{T}, p) where T
     TT = promote_op(^, T, typeof(p))
-    return (TT == T ? A : convert(AbstractMatrix{TT}, A))^Integer(p)
+    ATT = TT == T ? A : convert(AbstractMatrix{TT}, A)
+    return _integerpow(ATT, p)
 end
+_integerpow(A::AbstractMatrix, p) = A^Integer(p)
+function _integerpow(A::AbstractMatrix, p::Union{Float32, Float64})
+    # For these exponent types, not all values may be converted to Int
+    # We split the exponentiation into parts for which the exponent can be converted to an Int
+    if p < 0
+        return _integerpow(inv(A), -p)
+    end
+    m = typemax(Int)
+    if p <= m
```

Comment:
The flow of all of this seems a little more complicated than necessary (maybe I'm wrong). What about just computing `q, r = divrem(p, m)` initially and branching on `q > 0` rather than `p <= m`? It seems that could remove at least one `if` block.